x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/layer_actions.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { EditorFrameState } from './state_management';
+import { Datasource, Visualization } from '../../types';
+
+interface RemoveLayerOptions {
+  trackUiEvent: (name: string) => void;
+  state: EditorFrameState;
+  layerId: string;
+  activeVisualization: Pick<Visualization, 'getLayerIds' | 'clearLayer' | 'removeLayer'>;
+  datasourceMap: Record<string, Pick<Datasource, 'clearLayer' | 'removeLayer'>>;
+}
+
+interface AppendLayerOptions {
+  trackUiEvent: (name: string) => void;
+  state: EditorFrameState;
+  generateId: () => string;
+  activeDatasource: Pick<Datasource, 'insertLayer' | 'id'>;
+  activeVisualization: Pick<Visualization, 'appendLayer'>;
+}
+
+export function removeLayer(opts: RemoveLayerOptions): EditorFrameState {
+  const { state, trackUiEvent: trackUiEvent, activeVisualization, layerId, datasourceMap } = opts;
+  const isOnlyLayer = activeVisualization
+    .getLayerIds(state.visualization.state)
+    .every(id => id === opts.layerId);
+
+  trackUiEvent(isOnlyLayer ? 'layer_cleared' : 'layer_removed');
```

Comment:
I don't think trackUiEvent needs to be an argument, is this mainly for testing purposes? It's a singleton so it can be imported directly

Reply:
It's for testing, yeah. It's easier than mocking an import, that's all.